sklearn/linear_model/_base.py

```diff
@@ -287,11 +305,30 @@ def predict(self, X):
 
     def _set_intercept(self, X_offset, y_offset, X_scale):
         """Set the intercept_"""
+
+        xp, _ = get_namespace(X_offset, y_offset, X_scale)
+
         if self.fit_intercept:
             # We always want coef_.dtype=X.dtype. For instance, X.dtype can differ from
             # coef_.dtype if warm_start=True.
-            self.coef_ = np.divide(self.coef_, X_scale, dtype=X_scale.dtype)
-            self.intercept_ = y_offset - np.dot(X_offset, self.coef_.T)
+            coef_ = self.coef_ = xp.divide(
+                xp.astype(self.coef_, X_scale.dtype), X_scale
+            )
+            ravel = False
+            if self.coef_.ndim == 1:
+                coef_ = xp.reshape(coef_, (1, -1))
+                ravel = True
+
+            intercept_ = y_offset - (X_offset @ (coef_.T))
+
+            if ravel:
+                intercept_ = _ravel(intercept_, xp=xp)
+
+            if y_offset.ndim < 1:
+                intercept_ = intercept_[0]
+
+            self.intercept_ = intercept_
```

Comment:
I find this code with dependent condisions/reshaping/raveling very complex to reason about. Maybe we should instead have 2 branches based on `coef_.ndim` and do something like:

```python
if coef_.ndim == 2:
    self.intercept_ = y_offset - X_offset @ coef_.T
else:
    self.intercept_ = y_offset - X_offset @ coef_
```

Also: please no extra parens, especially around `coef.T`. I think the precedence of `@` over `+` is well known enough that parens around `X_offset @ coef_` are not needed either.